diff --git a/src/com/google/javascript/jscomp/parsing/JsDocInfoParser.java b/src/com/google/javascript/jscomp/parsing/JsDocInfoParser.java
index f1f57359..f2b037d2 100644
--- a/src/com/google/javascript/jscomp/parsing/JsDocInfoParser.java
+++ b/src/com/google/javascript/jscomp/parsing/JsDocInfoParser.java
@@ -1352,9 +1352,13 @@ public final class JsDocInfoParser {
 
     boolean ignoreStar = false;
 
+    int lineStartChar = -1;
     do {
       switch (token) {
         case STAR:
+          if (ignoreStar) {
+            lineStartChar = stream.getCharno() + 1;
+          }
           if (!ignoreStar) {
             if (builder.length() > 0) {
               builder.append(' ');
@@ -1372,6 +1376,7 @@ public final class JsDocInfoParser {
           }
 
           ignoreStar = true;
+          lineStartChar = 0;
           token = next();
           continue;
 
@@ -1380,8 +1385,25 @@ public final class JsDocInfoParser {
         case EOF:
           // When we're capturing a license block, annotations
           // in the block are ok.
-          if (!(option == WhitespaceOption.PRESERVE &&
-                token == JsDocToken.ANNOTATION)) {
+        default:
+          ignoreStar = false;
+          state = State.SEARCHING_ANNOTATION;
+          boolean isEOC = token == JsDocToken.EOC;
+          if (!isEOC) {
+            if (lineStartChar != -1 && option == WhitespaceOption.PRESERVE) {
+              int numSpaces = stream.getCharno() - lineStartChar;
+              for (int i = 0; i < numSpaces; i++) {
+                builder.append(' ');
+              }
+              lineStartChar = -1;
+            } else if (builder.length() > 0) {
+              builder.append(' ');
+            }
+          }
+          if (token == JsDocToken.EOC ||
+              token == JsDocToken.EOF ||
+              (token == JsDocToken.ANNOTATION &&
+               option != WhitespaceOption.PRESERVE)) {
             String multilineText = builder.toString();
 
             if (option != WhitespaceOption.PRESERVE) {
@@ -1401,13 +1423,6 @@ public final class JsDocInfoParser {
 
           // FALL THROUGH
 
-        default:
-          ignoreStar = false;
-          state = State.SEARCHING_ANNOTATION;
-
-          if (builder.length() > 0) {
-            builder.append(' ');
-          }
 
           builder.append(toString(token));
 
