diff --git a/src/com/google/javascript/jscomp/TypeInference.java b/src/com/google/javascript/jscomp/TypeInference.java
index bc248974..e2f462a2 100644
--- a/src/com/google/javascript/jscomp/TypeInference.java
+++ b/src/com/google/javascript/jscomp/TypeInference.java
@@ -1035,6 +1035,7 @@ class TypeInference
   private FlowScope traverseNew(Node n, FlowScope scope) {
     Node constructor = n.getFirstChild();
     scope = traverse(constructor, scope);
+    scope = traverseChildren(n, scope);
 
     JSType constructorType = constructor.getJSType();
     JSType type = null;
@@ -1052,14 +1053,12 @@ class TypeInference
         }
         if (ct != null && ct.isConstructor()) {
           type = ct.getInstanceType();
+          backwardsInferenceFromCallSite(n, ct);
         }
       }
     }
     n.setJSType(type);
 
-    for (Node arg = constructor.getNext(); arg != null; arg = arg.getNext()) {
-      scope = traverse(arg, scope);
-    }
     return scope;
   }
 
