diff --git a/src/main/java/org/apache/commons/lang3/time/FastDatePrinter.java b/src/main/java/org/apache/commons/lang3/time/FastDatePrinter.java
index cc1af9df..b2358600 100644
--- a/src/main/java/org/apache/commons/lang3/time/FastDatePrinter.java
+++ b/src/main/java/org/apache/commons/lang3/time/FastDatePrinter.java
@@ -1093,7 +1093,8 @@ public class FastDatePrinter implements DatePrinter, Serializable {
      * <p>Inner class to output a time zone name.</p>
      */
     private static class TimeZoneNameRule implements Rule {
-        private final TimeZone mTimeZone;
+        private final Locale mLocale;
+        private final int mStyle;
         private final String mStandard;
         private final String mDaylight;
 
@@ -1105,8 +1106,9 @@ public class FastDatePrinter implements DatePrinter, Serializable {
          * @param style the style
          */
         TimeZoneNameRule(TimeZone timeZone, Locale locale, int style) {
-            mTimeZone = timeZone;
 
+            mLocale = locale;
+            mStyle = style;
             mStandard = getTimeZoneDisplay(timeZone, false, style, locale);
             mDaylight = getTimeZoneDisplay(timeZone, true, style, locale);
         }
@@ -1124,10 +1126,11 @@ public class FastDatePrinter implements DatePrinter, Serializable {
          */
         @Override
         public void appendTo(StringBuffer buffer, Calendar calendar) {
-            if (mTimeZone.useDaylightTime() && calendar.get(Calendar.DST_OFFSET) != 0) {
+            TimeZone zone = calendar.getTimeZone();
+            if (zone.useDaylightTime() && calendar.get(Calendar.DST_OFFSET) != 0) {
                 buffer.append(mDaylight);
             } else {
-                buffer.append(mStandard);
+                buffer.append(getTimeZoneDisplay(zone, false, mStyle, mLocale));
             }
         }
     }
